Document/0x06h-Testing-Platform-Interaction.md

```diff
@@ -3,365 +3,2701 @@
+- Share: post to a sharing website or share content with others.
+- Today: also called widgets, they offer content or perform quick tasks in the Today view of Notification Center.
+
+###### How do app extensions interact with other apps?
+
+There are three important elements here:
+
+- App extension: is the one bundled inside a containing app. Host apps interact with it.
+- Host app: is the (third-party) app that triggers the app extension of another app.
+- Containing app: is the app that contains the app extension bundled into it.
+
+For example, the user selects text in the *host app*, clicks on the "Share" button and selects one "app" or action from the list. This triggers the *app extension* of the *containing app*. The app extension displays its view within the context of the host app and uses the items provided by the host app, the selected text in this case, to perform a specific task (post it on a social network, for example). See this picture from the [Apple App Extension Programming Guide](https://developer.apple.com/library/archive/documentation/General/Conceptual/ExtensibilityPG/ExtensionOverview.html#//apple_ref/doc/uid/TP40014214-CH2-SW13 "An app extension can communicate indirectly with its containing app") which pretty good summarizes this:
+
+![App Extensions Communication](Images/Chapters/0x06h/app_extensions_communication.png)
+
+###### Security Considerations
+
+From the security point of view it is important to note that
+
+- an app extension does never communicate directly with its containing app (typically, it isn’t even running while the contained app extension is running).
+- an app extension and the host app communicate via inter-process communication.
+- an app extension’s containing app and the host app don’t communicate at all.
+- a Today widget (and no other app extension type) can ask the system to open its containing app by calling the `openURL:completionHandler:` method of the `NSExtensionContext` class.
+- any app extension and its containing app can access shared data in a privately defined shared container.
+
+In addition:
+
+- app extensions cannot access some APIs, for example, HealthKit.
+- they cannot receive data using AirDrop but do can send data.
+- no long-running background tasks are allowed but uploads or downloads can be initiated.
+- app extensions cannot access the camera or microphone on an iOS device (except for iMessage app extensions).
+
+##### Static Analysis
+
+The static analysis will take care of:
+
+- Verifying if the App Contains App Extensions
+- Determining the Supported Data Types
+- Checking Data Sharing with the Containing App
+- Verifying if the App Restricts the Use of App Extensions
+
+###### Verifying if the App Contains App Extensions
+
+If you have the original source code you can search for all occurrences of `NSExtensionPointIdentifier` with Xcode (cmd+shift+f) or take a look into "Build Phases / Embed App extensions":
+
+![List App Extensions in Xcode](Images/Chapters/0x
```

Comment:
how can i make use of this? what is the danger? (e.g. explain what to do next with MiTMing given the mixed content and the possibility to do ssl bypass)

Reply:
is this really related to these lines? This section is about "Verifying if the App Contains App Extensions", it's simply an info-gathering step, so to say. The next step might be take a look into the app extension or reverse-engineer it. Apart from considering the following app extension specific steps:

- Determining the Supported Data Types
- Checking Data Sharing with the Containing App
- Verifying if the App Restricts the Use of App Extensions

Were you maybe referring here to the section "Testing for Mixed Content"?

Reply:
just to make sure, maybe i have missed it: but where are these steps tested?

Reply:
if you mean the ones from the bullet points, they are all in Static Analysis of the "#### App Extensions" section:

- Verifying if the App Contains App Extensions
- Determining the Supported Data Types
- Checking Data Sharing with the Containing App
- Verifying if the App Restricts the Use of App Extensions

Reply:
let's amke sure there is a separate issue for this: so that we can make the line of discussion a bit more clear and we can work towards merging this pr.